thirdparty/DiskANN/include/diskann/utils.h

```diff
@@ -642,24 +650,30 @@ namespace diskann {
       for (_u64 p = 0; p < block_pts; p++) {
         for (_u64 j = 0; j < in_dims; j++) {
           out_block_data[p * out_dims + j] =
-              in_block_data[p * in_dims + j] / max_norm;
+              (T) (((float) in_block_data[p * in_dims + j]) / max_norm);
         }
         float res = 1 - (norms[start_id + p] / (max_norm * max_norm));
         res = res <= 0 ? 0 : std::sqrt(res);
-        out_block_data[p * out_dims + out_dims - 1] = res;
+        out_block_data[p * out_dims + out_dims - 1] = (T) res;
       }
       out_writer.write((char*) out_block_data.get(),
-                       block_pts * out_dims * sizeof(float));
+                       block_pts * out_dims * sizeof(T));
     }
     out_writer.close();
     return max_norm;
   }
 
   template<typename T>
   std::vector<float> prepare_base_for_cosine(const std::string in_file,
-                                        const std::string out_file) {
-    LOG_KNOWHERE_DEBUG_
-        << "Pre-processing base file by normalizing";
+                                             const std::string out_file) {
+    if (!knowhere::KnowhereFloatTypeCheck<T>::value) {
+      std::stringstream stream;
+      stream
+          << "DiskANN currently only supports floating point data for Cosine."
```

Comment:
same note for `float`, `fp16` and `bf16`

Reply:
updated